core/src/saros/filesystem/ReferencePointComparator.java

```diff
@@ -0,0 +1,26 @@
+package saros.filesystem;
+
+import saros.SarosPluginContext;
+import saros.repackaged.picocontainer.annotations.Inject;
+
+/**
+ * Utility class providing static access to the {@link IReferencePointComparator} injected by the
+ * IDE-specific implementation.
+ */
+public class ReferencePointComparator {
+
+  @Inject private static IReferencePointComparator referencePointComparator;
+
+  static {
+    SarosPluginContext.initComponent(new ReferencePointComparator());
```

Comment:
I know there are several classes which are full of such hacks, but the way you use it among the SarosSession / ReferencePointMapper is evil. Basically it is spaghetti / static method code that is very hard to test.

As you have the container when the session is started you can retrieve it there instead of calling a static method.

Reply:
This class is ideal to offer static functionality. It is not dependent on any instance variables and offers general utility methods.

The issue is that the functionality has to be provided by the IDE-specific implementations, meaning the methods providing the functionality can't be static themselves. But having them be normal class and requesting the from the context if needed only further bloats up the constructors of our classes. A prime example for this is the path factory (or, as an extension, its Eclipse implementation). There is no need for it to be non-static besides it being injected by the IDE implementations. Its usage becomes so much more of a hassle just because it has to be requested from the context whenever it is needed. (I know that these classes aren't even used that much anymore. This is more meant as a general example.)

And while this class specifically can't be tested, that should not be an issue as it is only a surrogate to make accessing the actual implementations easier. This does not change anything about the testing behavior of the actual implementations. It slightly changes the testing behavior for testing classes that use the logic, but this can either be resolved by adding a mock to the context if it is used (which also needs to happen when testing the context without the static logic) or by mocking the static call using powermock.

Also, this setup would be a lot cleaner and far less "dirty"/"evil" if we were to add a getter for the plugin context (just like for the session context). This would allow us to avoid the reflection stuff and would replace the static block with just initializing the static variable using the getter. As this getter would also easily be mockable using powermock, stubbing the whole utility could be done using a test utility method.

Also, I don't get what the ReferencePointMapper has to do with this.

Reply:
The correct implementation regarding Unit testing of the core would be to implement a static util class in Eclipse, IntelliJ, and the Server. Afterwards you add just a IReferencePointComparer to the session context which just delegates to the static calls.

With your current implementation it is impossible to do a minimalistic instantiation of the SarosSession class with because of the static call to (I should have deleted this shit method) SarosPluginContext.initComponent class. It is simply not how DI works.

Reply:
When it comes to testing assume that mocking static methods is just a "bonus" and not the way to go.

In order to test this this I have to set a context. When I use the default context it starts creating an XMPP store etc. If I want to use another implementation I have to write my own initComponent method (which it btw marked as deprecated).

Can you please tell me in which places in the Core Annotation injection is used. We had an agreement to not use them in the Core at all, only CTOR injection.

Reply:
> When it comes to testing assume that mocking static methods is just a "bonus" and not the way to go.
> 
> In order to test this this I have to set a context. When I use the default context it starts creating an XMPP store etc. If I want to use another implementation I have to write my own initComponent method

Which is exactly what happens for `SarosSessionTest`. A general solution for this problem would be to create a test context factory that allows registering classes to mock (and optionally also which objects to use) and then creates a functioning mock context, including the injection logic.

> (which it btw marked as deprecated).

Yes, I noticed that as well. But the marking lacks any indication as to why and what to use instead. So I just assumed that this is something someone set at some point and then forgot about. Also, only the context method is marked as deprecated and not the static wrapper that is used to normally access it.

> Can you please tell me in which places in the Core Annotation injection is used.

As far as I can tell, the only class using annotation injection in the core (prior to this change) is `SarosSession` (see [variables](https://github.com/saros-project/saros/blob/master/core/src/saros/session/internal/SarosSession.java#L73) and [initialization](https://github.com/saros-project/saros/blob/master/core/src/saros/session/internal/SarosSession.java#L790)).

> We had an agreement to not use them in the Core at all, only CTOR injection.

I am not quite sure who "we" is in this context. I was not aware of any such agreement/rule and could not find any mention of it in the [new](https://www.saros-project.org/contribute/best-practices.html) or [old](https://saros-project.github.io/legacy_docs/coderules.1.html#dependency_injection) documentation.

If this is indeed something we want to avoid in the core, this is something we should definitely mention in the documentation. But, before we make such a decision, there should be a clear reasoning as to why.

In this discussion, it still has not become clear to me what the issue with annotation injection or its current usage is. Yes, testing gets harder, but not by that much in my opinion (but maybe I am already "numbed" by all the testing difficulties that arise with the usage of static IDE methods or other parts of the IDE context :sweat_smile:). Furthermore, the difficulties can be largely mitigated by introducing test helper classes.

Also, as [the old documentation already stated](https://saros-project.github.io/legacy_docs/coderules.1.html#dependency_injection), static access to the context is not always avoidable, e.g. when dealing with UI classes that are not initialized by Saros but by the IDE, so I don't think we can get rid of annotation injection (at least not without replacing it with another way of statically accessing the context).

Reply:
It all boils down regarding testing. Of course you can just use one argument. Every class will just get an IContext object and then calls getComponent to until the world collides. Just add a small taste of static massive calls that require even more setup, either through more static calls before calling the static call or even a more complex setup.

If you strictly follow this approach you will end up with classes that you simply do not want / or simply cannot test (and I have encountered that in my last job, you could not simply test one class in isolation, static method calls all over the place and there was no free of charge tool for C# that was able to mock static methods).

If you look at the SarosSessionTest, it is just a smoke test with an massive amount of setup. In reality it only checks if the forked PicoContainer is correctly detached and garbage collected.

We currently do not have any JUnit integration test(s) but only the STF. Of course in the STF everything is available but regarding JUnit integration tests and unit test every static method call that relies on other object that has to be properly initialized is just annoying.

As far as I remember the SarosPluginContext class was moved to the Core to avoid code duplication.

There was not really a discussion if X small line LOC classes in every plugin are more harmful than adding  this "EierlegendeWollmilchsau" to the Core and made it available to everything, including the Core.

It also do not really has something to do with Annotation Injection. That is just a matter of taste. In a real DI setup you throw everything into the DI framework and the classes are properly initialized.

However if you look at the implementation of the IContext#initComponent method it simply violates these principles.

To sum it up. You have the SarosSession class. It has full access to the the DI container. It creates a child container and get the factory for this child container from the DI context. Now after everything should be initialized you still have classes which will access static method calls which will start DI injection (lazy) again. To be more specific, if you call IContext#initComponent you are not part of the DI container. If such code is called from code that is part of a DI container then this is just bad design.

Reply:
Removed.